Refactor for upstream/master

mxmar opened the pull request.

blazejkrzak commented:
@mxmar @patred20  whats the status, what is the assignees/reviews milestones and stuff like this?

mxmar commented:
I'm changing PR owner. @patred20 Please open new PR with you as owner and assignee. Don't forget to set proper PR description with for e.g. change list and describe shortly PR reason. You can use this template:

```
## Description

[Describe what this change achieves]

### Related issue

[List any existing github issues this PR addresses]

### Change list
- added some functionality
- removed something
- refactored something

### Check List
- [ ] All tests pass.
- [ ] New functionality has been documented in the README if applicable
- [ ] Stable implementation
```
